drivers/display/ssd16xx.c

```diff
@@ -380,42 +381,73 @@ static int ssd16xx_set_window(const struct device *dev,
 	}
 
-	if ((y + desc->height) > panel_h) {
-		LOG_ERR("Buffer out of bounds (height)");
-		return -EINVAL;
-	}
+	if (data->orientation == DISPLAY_ORIENTATION_NORMAL ||
+	    data->orientation == DISPLAY_ORIENTATION_ROTATED_180) {
+		if ((y + desc->height) > panel_h) {
+			LOG_ERR("Buffer out of bounds (height)");
+			return -EINVAL;
+		}
 
-	if ((x + desc->width) > config->width) {
-		LOG_ERR("Buffer out of bounds (width)");
-		return -EINVAL;
-	}
+		if ((x + desc->width) > config->width) {
+			LOG_ERR("Buffer out of bounds (width)");
+			return -EINVAL;
+		}
 
-	if ((desc->height % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0U) {
-		LOG_ERR("Buffer height not multiple of %d",
-				EPD_PANEL_NUMOF_ROWS_PER_PAGE);
-		return -EINVAL;
-	}
+		if ((desc->height % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0U) {
+			LOG_ERR("Buffer height not multiple of %d", EPD_PANEL_NUMOF_ROWS_PER_PAGE);
+			return -EINVAL;
+		}
 
-	if ((y % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0U) {
-		LOG_ERR("Y coordinate not multiple of %d",
-				EPD_PANEL_NUMOF_ROWS_PER_PAGE);
-		return -EINVAL;
+		if ((y % EPD_PANEL_NUMOF_ROWS_PER_PAGE) != 0U) {
+			LOG_ERR("Y coordinate not multiple of %d", EPD_PANEL_NUMOF_ROWS_PER_PAGE);
+			return -EINVAL;
+		}
+	} else {
+		if ((y + desc->height) > config->width) {
+			LOG_ERR("Buffer out of bounds (height)");
+			return -EINVAL;
+		}
+
+		if ((x + desc->width) > panel_h) {
+			LOG_ERR("Buffer out of bounds (width)");
+			return -EINVAL;
+		}
+
+		if ((desc->width % SSD16XX_PIXELS_PER_BYTE) != 0U) {
+			LOG_ERR("Buffer width not multiple of %d", SSD16XX_PIXELS_PER_BYTE);
+			return -EINVAL;
+		}
+
+		if ((x % SSD16XX_PIXELS_PER_BYTE) != 0U) {
+			LOG_ERR("X coordinate not multiple of %d", SSD16XX_PIXELS_PER_BYTE);
+			return -EINVAL;
+		}
 	}
 
-	switch (data->scan_mode) {
-	case SSD16XX_DATA_ENTRY_XIYDY:
+	switch (data->orientation) {
+	case DISPLAY_ORIENTATION_NORMAL:
+		x_start = (panel_h - 1 - y) / SSD16XX_PIXELS_PER_BYTE;
+		x_end = (panel_h - 1 - (y + desc->height - 1)) / SSD16XX_PIXELS_PER_BYTE;
+		y_start = x;
+		y_end = (x + desc->width - 1);
+		break;
+	case DISPLAY_ORIENTATION_ROTATED_90:
+		x_start = (panel_h - 1 - x) / SSD16XX_PIXELS_PER_BYTE;
+		x_end = (panel_h - 1 - (x + desc->width - 1)) / SSD16XX_PIXELS_PER_BYTE;
+		y_start = (config->width - 1 - y);
+		y_end = (config->width - 1 - (y + desc->height - 1));
+		break;
+	case DISPLAY_ORIENTATION_ROTATED_180:
 		x_start = y / SSD16XX_PIXELS_PER_BYTE;
 		x_end = (y + desc->height - 1) / SSD16XX_PIXELS_PER_BYTE;
 		y_start = (x + desc->width - 1);
 		y_end = x;
 		break;
-
-	case SSD16XX_DATA_ENTRY_XDYIY:
-		x_start = (panel_h - 1 - y) / SSD16XX_PIXELS_PER_BYTE;
-		x_end = (panel_h - 1 - (y + desc->height - 1)) /
-			SSD16XX_PIXELS_PER_BYTE;
-		y_start = x;
-		y_end = (x + desc->width - 1);
+	case DISPLAY_ORIENTATION_ROTATED_270:
+		x_start = x / SSD16XX_PIXELS_PER_BYTE;
+		x_end = (x + desc->width - 1) / SSD16XX_PIXELS_PER_BYTE;
+		y_start = y;
+		y_end = (y + desc->height - 1);
```

Comment:
Do we still need to write the `scan_mode` here, since we do it within the `set_orientation` function? IE will the scan mode setting persist between display frames, or does it need to be set every time?

Reply:
Thanks for reviewing @danieldegrasse!

In the second commit I removed writing the `scan_mode` on each `set_window` call because I didn't find any indication in the datasheet that this should not persist between resets. So I tested it with the two mentioned display controllers and it worked so that seems to prove that the `scan_mode` indeed persists between display frames. 
Because there is a software reset in the `set_profile` function I also added writing the `scan_mode` over there.

I do wonder why it was added to the `set_window` function in the first place instead of in the `controller_init` so I completely understand your doubts.
If anyone with a ssd16xx laying around could double check this, that would be great.
@andysan, I see you have made some amazing improvements to this driver in the past, would you mind checking this as well?